sdk/iot/provisioning/src/az_iot_provisioning_client.c

```diff
@@ -47,161 +54,420 @@ AZ_NODISCARD az_result az_iot_provisioning_client_init(
+    size_t* out_mqtt_topic_filter_length)
 {
   (void)client;
 
   AZ_PRECONDITION_NOT_NULL(client);
-  AZ_PRECONDITION_VALID_SPAN(mqtt_topic_filter, 0, false);
-  AZ_PRECONDITION_NOT_NULL(out_mqtt_topic_filter);
+  AZ_PRECONDITION_NOT_NULL(mqtt_topic_filter);
+  AZ_PRECONDITION(mqtt_topic_filter_size > 0);
 
-  int32_t required_length
-      = az_span_size(str_dps) + az_span_size(str_registrations) + az_span_size(str_res) + 1;
+  az_span mqtt_topic_filter_span
+      = az_span_init((uint8_t*)mqtt_topic_filter, (int32_t)mqtt_topic_filter_size);
+  int32_t required_length = az_span_size(str_dps_registrations_res) + 1;
 
-  AZ_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_topic_filter, required_length);
+  AZ_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_topic_filter_span, required_length);
 
-  az_span remainder = az_span_copy(mqtt_topic_filter, str_dps);
-  remainder = az_span_copy(remainder, str_registrations);
-  remainder = az_span_copy(remainder, str_res);
-  az_span_copy_u8(remainder, '#');
+  az_span remainder = az_span_copy(mqtt_topic_filter_span, str_dps_registrations_res);
+  remainder = az_span_copy_u8(remainder, '#');
+  remainder = az_span_copy_u8(remainder, '\0');
 
-  *out_mqtt_topic_filter = az_span_slice(mqtt_topic_filter, 0, required_length);
+  if (out_mqtt_topic_filter_length)
+  {
+    *out_mqtt_topic_filter_length = (size_t)required_length;
+  }
 
   return AZ_OK;
 }
 
-// topic: $dps/registrations/res/202/?$rid=1&retry-after=3
-// payload:
-// {"operationId":"4.d0a671905ea5b2c8.42d78160-4c78-479e-8be7-61d5e55dac0d","status":"assigning"}
-AZ_NODISCARD az_result az_iot_provisioning_client_received_topic_payload_parse(
+AZ_INLINE az_iot_provisioning_client_registration_state registration_state_default()
+{
+  return (az_iot_provisioning_client_registration_state){ .assigned_hub_hostname = AZ_SPAN_NULL,
+                                                          .device_id = AZ_SPAN_NULL,
+                                                          .status = AZ_IOT_STATUS_OK,
+                                                          .extended_error_code = 0,
+                                                          .error_message = AZ_SPAN_NULL,
+                                                          .error_tracking_id = AZ_SPAN_NULL,
+                                                          .error_timestamp = AZ_SPAN_NULL };
+}
+
+/*
+Documented at
+https://docs.microsoft.com/en-us/rest/api/iot-dps/runtimeregistration/registerdevice#deviceregistrationresult
+  "registrationState":{
+    "x509":{},
+    "registrationId":"paho-sample-device1",
+    "createdDateTimeUtc":"2020-04-10T03:11:13.0276997Z",
+    "assignedHub":"contoso.azure-devices.net",
+    "deviceId":"paho-sample-device1",
+    "status":"assigned",
+    "substatus":"initialAssignment",
+    "lastUpdatedDateTimeUtc":"2020-04-10T03:11:13.2096201Z",
+    "etag":"IjYxMDA4ZDQ2LTAwMDAtMDEwMC0wMDAwLTVlOGZlM2QxMDAwMCI="}}
+*/
+AZ_INLINE az_result az_iot_provisioning_client_payload_error_code_parse(
```

Comment:
should these parse methods be this instead to follow the changed "get" methods
```suggestion
AZ_INLINE az_result az_iot_provisioning_client_parse_payload_error_code(
```

Reply:
I forgot to review these since they were internal. Makes sense.